# \(More\) Assembly QC

Joe Fass - jnfass@ucdavis.edu  
14 December 2016

## N50 and Related (Better!) Metrics

Test NG50; list sizes, add cumulative length column, print at desired cumulative length cutoff  

```bash
cat p_ctg.fa | \
    ./fa.sizes.pl | sort -rn | \
    perl -ne 'chomp; $c+=$_; print "$_\t$c\n"' | \
    perl -ane 'if ($F[1] >= 30000000) {print; exit}'
```

List base content:  

```bash
cat p_ctg.fa | grep -v ">" | grep -o . | \
    perl -ne 'chomp; $h{$_}++ }{ print "$_\t$h{$_}\n" for (keys %h)'
```



---

## BUSCO

First download a relevant lineage tarball from [Evgeny Zdobnov's lab](http://busco.ezlab.org/).

```bash
tar xzvf fungi_odb9.tar.gz
```

We don't have a module for the most recent BUSCO yet, but the old module sets up the dependencies. So, load the module, but use the newer BUSCO script. Also we need to have a local Augustus config directory, so BUSCO has permission to muck around in it and make changes.

```bash
module load busco/1.1b1
# download augustus-3.0.3 locally, to point at legitimate config directory
wget http://bioinf.uni-greifswald.de/augustus/binaries/old/augustus-3.0.3.tar.gz
tar xzvf augustus-3.0.3.tar.gz
export AUGUSTUS_CONFIG_PATH=/share/biocore/jfass/Bioinformatics-Genome-Assembly-and-Analysis-Workshop-Pac-Bio-and-10x-Genomics-/augustus-3.0.3/config/
# download BUSCO 2.0 locally to use updated script
./busco-master-623f5a65e8467daea4dd1eedf13653a52de25897/BUSCO.py \
    --in p_ctg.fa --out pet_fungus_BUSCO --lineage \
    /share/biocore/jfass/Bioinformatics-Genome-Assembly-and-Analysis-Workshop-Pac-Bio-and-10x-Genomics-/fungi_odb9 \
    --mode genome -f
```

## Variant viz (?)

Now how to address variants found by falcon\_unzip (found in 4-quiver/cns\_output/)? Try mauve-contig-reorder on all #F\_### contigs collected together (haplotigs) versus collected (or collected and reordered) primary contigs. Alternately separate haplotigs by each primary 'tig they align to (use LAST, or Zev's modified BLASR), to align each separately.  

```bash
tar xzvf mauve_linux_snapshot_2015-02-13.tar.gz
cd mauve_snapshot_2015-02-13/
srun java -Xmx5g -cp Mauve.jar org.gel.mauve.contigs.ContigOrderer -output ../mauve.reorder -ref ../cns_p_ctg.fasta -draft ../cns_h_ctg.fasta
cd ..
```

Pay attention to the last alignment folder (could be any number) generated by Mauve; it contains the final reordered set of contigs named the same (or sometimes with an extra ".fas" extension). We can see that the order of contigs has changed:  

```bash
cat cns_h_ctg.fasta | grep ">" | head
cat mauve.reorder/alignment6/cns_h_ctg.fasta | grep ">" | head
```

We'll need to pull the alignment files down to a local machine to view them.  

```bash
cd mauve.reorder/alignment6/
cp alignment6 alignment6.original
# then open a terminal on your local machine, and:
rsync -avzn class8@cabernet.genomecenter.ucdavis.edu:/share/biocore/workshops/class8/mauve.reorder/alignment6 .
# remove "n" from rsync's options once you're satisfied you're grabbing the right directory
```

Inspect the alignment by launching Mauve, then File --> Open alignment:  

```bash
wget http://darlinglab.org/mauve/snapshots/2015/2015-02-13/linux-x64/mauve_linux_snapshot_2015-02-13.tar.gz
tar xzvf mauve_linux_snapshot_2015-02-13.tar.gz
cd mauve_snapshot_2015-02-13/
java -Xmx1g -jar Mauve.jar
```

Back on the server, pull out the cumulative lengths, as above, and plot at will. We can pull out the sizes of the alignments ... in this case, they represent the fraction of the genome that contains structural variation, according to FALCON\_unzip. Look in the backbone file. Pull out the sizes of alignment blocks, and cumulative lengths, this way: 

```bash
tail -n +2 mauve.reorder/alignment6/alignment6.backbone | cut -f1-2 | \
    perl -ane '$b=abs($F[1]-$F[0]); print "$b\n"' | \
    sort -rn | perl -ne '$b=$_; chomp $b; $c+=$b; \
    print "$b\t$c\n"' > blocks.lengths
gnuplot
set term dumb
plot "blocks.lengths" with points
plot [1:5e6] [1:7e7] "blocks.lengths" with points
# Ctrl-d to quit
```





Possible section assigning haplotigs to primotigs

```bash
module load last/621
lastdb primotigs.lastdb cns_p_ctg.fasta

```
